File: packages/python/src/a2a_lite/push_notifications.py
"""
Push notification support for A2A Lite agents.

Usage (agent-level):
    from a2a_lite.push_notifications import WebhookPushNotifier

    agent = Agent(
        name="Bot",
        description="...",
        push_notifier=WebhookPushNotifier(
            url="https://my-app.com/webhook/a2a",
            secret="my-signing-secret",  # optional HMAC-SHA256 signing
        ),
    )

Usage (per-task):
    from a2a_lite.push_notifications import TaskPushRegistry

    # Server-side: registry is auto-created on every Agent.
    # Client-side: use set_task_push_notification() or handle.subscribe().
"""

from __future__ import annotations

import hashlib
import hmac
import json
import logging
from abc import ABC, abstractmethod
from typing import Any

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Per-task push notification registry & middleware
# ---------------------------------------------------------------------------


class TaskPushRegistry:
    """
    Server-side registry mapping task IDs to webhook configurations.
    Used to deliver per-task push notifications when tasks complete.
    """

    def __init__(self) -> None:
        self._configs: dict[str, dict] = {}  # task_id -> {url, token}

    def set(self, task_id: str, url: str, token: str | None = None) -> None:
        self._configs[task_id] = {"url": url, "token": token}

    def get(self, task_id: str) -> dict | None:
        return self._configs.get(task_id)

    def delete(self, task_id: str) -> bool:
        return self._configs.pop(task_id, None) is not None

    def __contains__(self, task_id: str) -> bool:
        return task_id in self._configs


class PushNotificationMiddleware:
    """
    Starlette ASGI middleware that handles the A2A v1.0 push notification
    JSON-RPC methods (CreateTaskPushNotificationConfig, GetTaskPushNotificationConfig,
    DeleteTaskPushNotificationConfig).
    Wraps the underlying A2A SDK app and intercepts these specific methods.
    """

    def __init__(self, app: Any, registry: TaskPushRegistry) -> None:
        self.app = app
        self.registry = registry

    def __getattr__(self, name: str) -> Any:
        # Proxy attribute access to the wrapped app so callers can still
        # inspect Starlette-specific attributes (e.g. ``routes``).
        return getattr(self.app, name)

    async def __call__(self, scope: dict, receive: Any, send: Any) -> None:
        if scope["type"] != "http" or scope.get("method") != "POST":
            await self.app(scope, receive, send)
            return

        # Buffer the request body so we can read it and also replay it
        body_parts: list[bytes] = []
        more_body = True
        while more_body:
            message = await receive()
            body_parts.append(message.get("body", b""))
            more_body = message.get("more_body", False)
        body = b"".join(body_parts)

        try:
            data = json.loads(body)
            method = data.get("method", "")
        except (json.JSONDecodeError, AttributeError):
            method = ""

        if method == "CreateTaskPushNotificationConfig":
            await self._handle_set(data, send)
        elif method == "GetTaskPushNotificationConfig":
            await self._handle_get(data, send)
        elif method == "DeleteTaskPushNotificationConfig":
            await self._handle_delete(data, send)
        else:
            # Not a push notification method -- replay body to original app
            replayed = False

            async def replay_receive() -> dict:
                nonlocal replayed
                if not replayed:
                    replayed = True
                    return {"type": "http.request", "body": body, "more_body": False}
                # Pass through so SSE responses can still detect real disconnects
                return await receive()

            await self.app(scope, replay_receive, send)

    async def _send_json(self, send: Any, data: dict, status: int = 200) -> None:
        body = json.dumps(data).encode()
        await send(
            {
                "type": "http.response.start",
                "status": status,
                "headers": [
                    (b"content-type", b"application/json"),
                    (b"content-length", str(len(body)).encode()),
                ],
            }
        )
        await send({"type": "http.response.body", "body": body})

    async def _handle_set(self, data: dict, send: Any) -> None:
        params = data.get("params", {})
        task_id = params.get("taskId", "")
        config_id = params.get("id") or task_id
        url = params.get("url", "")
        token = params.get("token")
        if not task_id or not url:
            await self._send_json(
                send,
                {
                    "jsonrpc": "2.0",
                    "id": data.get("id"),
                    "error": {"code": -32602, "message": "Missing required fields: taskId and url"},
                },
            )
            return
        self.registry.set(task_id, url, token)
        await self._send_json(
            send,
            {
                "jsonrpc": "2.0",
                "id": data.get("id"),
                "result": {"taskId": task_id, "id": config_id, "url": url, "token": token},
            },
        )

    async def _handle_get(self, data: dict, send: Any) -> None:
        params = data.get("params", {})
        task_id = params.get("taskId", "")
        config_id = params.get("id") or task_id
        config = self.registry.get(task_id)
        if config is None:
            await self._send_json(
                send,
                {
                    "jsonrpc": "2.0",
                    "id": data.get("id"),
                    "error": {"code": -32001, "message": f"No push notification config for task {task_id}"},
                },
            )
            return
        await self._send_json(
            send,
            {
                "jsonrpc": "2.0",
                "id": data.get("id"),
                "result": {"taskId": task_id, "id": config_id, "url": config["url"], "token": config["token"]},
            },
        )

    async def _handle_delete(self, data: dict, send: Any) -> None:
        params = data.get("params", {})
        task_id = params.get("taskId", "")
        self.registry.delete(task_id)
        await self._send_json(
            send,
            {
                "jsonrpc": "2.0",
                "id": data.get("id"),
                "result": {},
            },
        )


class PushNotifier(ABC):
    """
    Abstract base class for push notification delivery.

    Implement this to send task completion events to external systems
    (webhooks, message queues, Slack, etc.).

    Example custom implementation:
        class SlackPushNotifier(PushNotifier):
            def __init__(self, webhook_url: str):
                self.url = webhook_url

            async def notify(self, event: dict) -> None:
                import httpx
                text = f"Skill `{event['skill']}` completed: {event['status']}"
                async with httpx.AsyncClient() as client:
                    await client.post(self.url, json={"text": text})
    """

    @abstractmethod
    async def notify(self, event: dict[str, Any]) -> None:
        """
        Send a notification for a skill completion event.

        Args:
            event: dict with keys:
                - skill (str): the skill that completed
                - result (Any): the skill's return value
                - status (str): "completed" | "failed"
                - timestamp (float): unix timestamp
                - agent (str): agent name
        """
        ...


class WebhookPushNotifier(PushNotifier):
    """
    Sends skill completion events as HTTP POST requests to a webhook URL.

    Features:
    - Automatic retry with exponential backoff
    - Optional HMAC-SHA256 request signing
    - Configurable headers
    - Timeout protection

    Example:
        notifier = WebhookPushNotifier(
            url="https://api.example.com/a2a-events",
            secret="my-webhook-secret",
            headers={"Authorization": "Bearer token"},
            max_retries=3,
        )
    """

    def __init__(
        self,
        url: str,
        secret: str | None = None,
        headers: dict[str, str] | None = None,
        max_retries: int = 3,
        timeout: float = 10.0,
    ) -> None:
        """
        Args:
            url: Webhook endpoint URL
            secret: Optional secret for HMAC-SHA256 request signing.
                    The signature is sent as X-A2A-Signature header.
            headers: Additional HTTP headers to include in requests.
            max_retries: Number of retry attempts on failure (default: 3).
            timeout: HTTP request timeout in seconds (default: 10.0).
        """
        self.url = url
        self.secret = secret
        self.headers = headers or {}
        self.max_retries = max_retries
        self.timeout = timeout

    def _sign_payload(self, payload: str) -> str:
        """Generate HMAC-SHA256 signature for the payload."""
        return hmac.new(
            self.secret.encode("utf-8"),
            payload.encode("utf-8"),
            hashlib.sha256,
        ).hexdigest()

    async def notify(self, event: dict[str, Any]) -> None:
        """Send the event to the configured webhook URL with retries."""
        try:
            import httpx
        except ImportError:
            logger.error("httpx is required for WebhookPushNotifier. Install it with: pip install httpx")
            return

        payload = json.dumps(event, default=str)
        headers = {
            "Content-Type": "application/json",
            "X-A2A-Event": event.get("skill", "unknown"),
            **self.headers,
        }

        if self.secret:
            headers["X-A2A-Signature"] = f"sha256={self._sign_payload(payload)}"

        last_error: Exception | None = None
        for attempt in range(self.max_retries):
            try:
                async with httpx.AsyncClient() as client:
                    response = await client.post(
                        self.url,
                        content=payload,
                        headers=headers,
                        timeout=self.timeout,
                    )
                    response.raise_for_status()
                    logger.debug(
                        "Push notification sent: skill=%s status=%s",
                        event.get("skill"),
                        response.status_code,
                    )
                    return
            except Exception as e:
                last_error = e
                if attempt < self.max_retries - 1:
                    wait = 2**attempt  # 1s, 2s, 4s
                    logger.warning(
                        "Push notification failed (attempt %d/%d), retrying in %ds: %s",
                        attempt + 1,
                        self.max_retries,
                        wait,
                        e,
                    )
                    import asyncio

                    await asyncio.sleep(wait)

        logger.error(
            "Push notification failed after %d attempts: %s",
            self.max_retries,
            last_error,
        )


class LogPushNotifier(PushNotifier):
    """
    Simple notifier that logs events. Useful for development and debugging.

    Example:
        agent = Agent(
            name="Bot",
            description="...",
            push_notifier=LogPushNotifier(level="INFO"),
        )
    """

    def __init__(self, level: str = "INFO") -> None:
        self.level = level.upper()

    async def notify(self, event: dict[str, Any]) -> None:
        log_fn = getattr(logger, self.level.lower(), logger.info)
        log_fn(
            "[A2A Push] skill=%s status=%s timestamp=%s",
            event.get("skill"),
            event.get("status"),
            event.get("timestamp"),
        )

File: packages/typescript/src/push-notifications.ts
/**
 * Push notification support for A2A Lite agents.
 *
 * @example
 * import { Agent } from './agent';
 * import { WebhookPushNotifier } from './push-notifications';
 *
 * const agent = new Agent({
 *   name: 'Bot',
 *   description: '...',
 *   pushNotifier: new WebhookPushNotifier({
 *     url: 'https://my-app.com/webhook/a2a',
 *     secret: 'my-signing-secret',
 *   }),
 * });
 */

import { createHmac } from 'crypto';

/** Event payload sent to push notifiers on skill completion. */
export interface PushEvent {
  skill: string;
  result: unknown;
  status: 'completed' | 'failed';
  timestamp: number;
  agent: string;
}

/**
 * Abstract base class for push notification delivery.
 * Extend this to send events to Slack, queues, webhooks, etc.
 *
 * @example
 * class SlackPushNotifier extends PushNotifier {
 *   async notify(event: PushEvent): Promise<void> {
 *     await fetch(this.slackUrl, {
 *       method: 'POST',
 *       body: JSON.stringify({ text: `${event.skill} completed` }),
 *     });
 *   }
 * }
 */
export abstract class PushNotifier {
  abstract notify(event: PushEvent): Promise<void>;
}

export interface WebhookPushNotifierOptions {
  /** Webhook endpoint URL */
  url: string;
  /**
   * Optional secret for HMAC-SHA256 request signing.
   * Signature sent as X-A2A-Signature header: `sha256=<hex>`
   */
  secret?: string;
  /** Additional HTTP headers */
  headers?: Record<string, string>;
  /** Number of retry attempts on failure (default: 3) */
  maxRetries?: number;
  /** Request timeout in milliseconds (default: 10000) */
  timeoutMs?: number;
}

/**
 * Sends skill completion events as HTTP POST requests to a webhook URL.
 *
 * Features:
 * - Automatic retry with exponential backoff
 * - Optional HMAC-SHA256 request signing
 * - Configurable headers and timeout
 *
 * @example
 * const notifier = new WebhookPushNotifier({
 *   url: 'https://api.example.com/a2a-events',
 *   secret: 'my-webhook-secret',
 *   headers: { 'Authorization': 'Bearer token' },
 *   maxRetries: 3,
 * });
 */
export class WebhookPushNotifier extends PushNotifier {
  private readonly url: string;
  private readonly secret?: string;
  private readonly headers: Record<string, string>;
  private readonly maxRetries: number;
  private readonly timeoutMs: number;

  constructor(options: WebhookPushNotifierOptions) {
    super();
    this.url = options.url;
    this.secret = options.secret;
    this.headers = options.headers ?? {};
    this.maxRetries = options.maxRetries ?? 3;
    this.timeoutMs = options.timeoutMs ?? 10_000;
  }

  private sign(payload: string): string {
    return `sha256=${createHmac('sha256', this.secret!).update(payload).digest('hex')}`;
  }

  async notify(event: PushEvent): Promise<void> {
    const payload = JSON.stringify(event);
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
      'X-A2A-Event': event.skill,
      ...this.headers,
    };

    if (this.secret) {
      headers['X-A2A-Signature'] = this.sign(payload);
    }

    let lastError: unknown;

    for (let attempt = 0; attempt < this.maxRetries; attempt++) {
      try {
        const controller = new AbortController();
        const timer = setTimeout(() => controller.abort(), this.timeoutMs);

        try {
          const response = await fetch(this.url, {
            method: 'POST',
            body: payload,
            headers,
            signal: controller.signal,
          });

          if (!response.ok) {
            throw new Error(`Webhook responded with ${response.status}`);
          }
          return; // success
        } finally {
          clearTimeout(timer);
        }
      } catch (err) {
        lastError = err;
        if (attempt < this.maxRetries - 1) {
          const waitMs = 1000 * Math.pow(2, attempt); // 1s, 2s, 4s
          await new Promise((resolve) => setTimeout(resolve, waitMs));
        }
      }
    }

    console.error(`[A2A] Push notification failed after ${this.maxRetries} attempts:`, lastError);
  }
}

/**
 * Development notifier that logs events to console.
 * Useful for testing push notification wiring without a real endpoint.
 *
 * @example
 * const agent = new Agent({
 *   name: 'Bot',
 *   description: '...',
 *   pushNotifier: new LogPushNotifier(),
 * });
 */
// ---------------------------------------------------------------------------
// TaskPushRegistry — per-task push notification config store
// ---------------------------------------------------------------------------

/**
 * In-memory registry for per-task push notification configurations.
 *
 * A caller registers a webhook URL (and optional bearer token) for a specific
 * task ID via the `CreateTaskPushNotificationConfig` JSON-RPC method.  When the
 * task completes, the server fires the webhook.
 */
export class TaskPushRegistry {
  private readonly configs = new Map<string, { url: string; token?: string }>();

  set(taskId: string, url: string, token?: string): void {
    this.configs.set(taskId, { url, token });
  }

  get(taskId: string): { url: string; token?: string } | undefined {
    return this.configs.get(taskId);
  }

  delete(taskId: string): boolean {
    return this.configs.delete(taskId);
  }

  has(taskId: string): boolean {
    return this.configs.has(taskId);
  }
}

// ---------------------------------------------------------------------------
// createPushNotificationMiddleware — Express middleware for JSON-RPC methods
// ---------------------------------------------------------------------------

/**
 * Returns Express middleware that intercepts the A2A v1.0 push notification
 * JSON-RPC methods (`CreateTaskPushNotificationConfig`, `GetTaskPushNotificationConfig`,
 * `DeleteTaskPushNotificationConfig`) and manages the per-task registry.
 *
 * Mount this **before** the SDK's `jsonRpcHandler` so that the custom
 * methods are handled without reaching the protocol layer.
 */
export function createPushNotificationMiddleware(registry: TaskPushRegistry) {
  return async (req: any, res: any, next: () => void): Promise<void> => {
    if (req.method !== 'POST' || !req.body) {
      return next();
    }

    const { method, id, params } = req.body;

    if (method === 'CreateTaskPushNotificationConfig') {
      const taskId = params?.taskId;
      const configId = params?.id ?? taskId;
      const url = params?.url;
      const token = params?.token;

      if (!taskId || !url) {
        res.json({
          jsonrpc: '2.0',
          id: id ?? null,
          error: { code: -32602, message: 'Missing required fields: taskId and url' },
        });
        return;
      }

      registry.set(taskId, url, token);

      res.json({
        jsonrpc: '2.0',
        id: id ?? null,
        result: { taskId, id: configId, url, token },
      });
      return;
    }

    if (method === 'GetTaskPushNotificationConfig') {
      const taskId = params?.taskId;
      const configId = params?.id ?? taskId;

      if (!taskId) {
        res.json({
          jsonrpc: '2.0',
          id: id ?? null,
          error: { code: -32602, message: 'Missing taskId' },
        });
        return;
      }

      const config = registry.get(taskId);
      if (!config) {
        res.json({
          jsonrpc: '2.0',
          id: id ?? null,
          error: { code: -32001, message: `No push notification config for task ${taskId}` },
        });
        return;
      }

      res.json({
        jsonrpc: '2.0',
        id: id ?? null,
        result: { taskId, id: configId, url: config.url, token: config.token },
      });
      return;
    }

    if (method === 'DeleteTaskPushNotificationConfig') {
      const taskId = params?.taskId;

      if (!taskId) {
        res.json({
          jsonrpc: '2.0',
          id: id ?? null,
          error: { code: -32602, message: 'Missing taskId' },
        });
        return;
      }

      registry.delete(taskId);
      res.json({
        jsonrpc: '2.0',
        id: id ?? null,
        result: {},
      });
      return;
    }

    return next();
  };
}

// ---------------------------------------------------------------------------
// LogPushNotifier
// ---------------------------------------------------------------------------

export class LogPushNotifier extends PushNotifier {
  async notify(event: PushEvent): Promise<void> {
    console.log('[A2A Push]', {
      skill: event.skill,
      status: event.status,
      agent: event.agent,
      timestamp: new Date(event.timestamp * 1000).toISOString(),
    });
  }
}
